File: CoreLib/src/main/java/org/hexnibble/corelib/commands/rc/group/ParallelRC.java
package org.hexnibble.corelib.commands.rc.group;

import org.hexnibble.corelib.commands.rc.RC;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

/**
 * Run multiple commands in parallel, ending only when all commands are completed.
 */
public class ParallelRC extends RC {
   private final List<RC> commandList;

   public ParallelRC(List<RC> commandList) {
      super("ParallelRC");
      this.commandList = new ArrayList<>(commandList);
   }

   public ParallelRC(RC... commands) {
      this(Arrays.asList(commands));
   }

   @Override
   protected void processCommand() {
      // removeIf will cycle through each command in the List and remove it if it is complete
      commandList.removeIf(RC::processRC);
      if (commandList.isEmpty()) {
         setCommandStatus(COMMAND_STATUS.COMMAND_SUCCESSFULLY_COMPLETED);
      }
   }
}

File: CoreLib/src/main/java/org/hexnibble/corelib/commands/rc/ParallelRC.kt
package org.hexnibble.corelib.commands.rc

/**
 * Run multiple commands in parallel, ending as soon as a single command completes.
 * @param commandList The list of RCs to run.
 * @param commandID The ID of the command.
 * @see ParallelRaceRC
 */
class ParallelRC @JvmOverloads constructor(
  private var commandList: MutableList<RC>,
  commandID: String = "ParallelRC"
) : RC(commandID) {
  /**
   * Run multiple commands in parallel, ending as soon as a single command completes.
   * @param commandList The RCs to run
   * @param commandID The ID of the command
   * @see ParallelRaceRC
   */
  @JvmOverloads
  constructor(vararg commandList: RC, commandID: String = "ParallelRC") : this(
    mutableListOf(*commandList),
    commandID
  )

  override fun processCommand() {
    // Remove command if complete
    commandList.removeIf { obj: RC -> obj.processRC() }
    if (commandList.isEmpty()) {
      commandStatus = COMMAND_STATUSES.COMMAND_SUCCESSFULLY_COMPLETED
    }
  }
}